File: android/src/main/kotlin/in/appamania/envified/EnvifiedPlugin.kt
package `in`.appamania.envified

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONArray
import java.security.SecureRandom

/**
 * Flutter plugin entry point for envified.
 *
 * Dispatches method-channel calls to [EnvifiedKeystore], [EnvifiedCrypto],
 * and the encrypted-prefs storage layer.
 */
class EnvifiedPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    // ── FlutterPlugin ────────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // ── MethodCallHandler ────────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                "initialize" -> handleInitialize(call, result)
                "encrypt" -> handleEncrypt(call, result)
                "decrypt" -> handleDecrypt(call, result)
                "storeSecret" -> handleStoreSecret(call, result)
                "retrieveSecret" -> handleRetrieveSecret(call, result)
                "deleteSecret" -> handleDeleteSecret(call, result)
                "deleteAllSecrets" -> handleDeleteAllSecrets(call, result)
                "keyExists" -> handleKeyExists(call, result)
                "rotateKey" -> handleRotateKey(call, result)
                "persistConfig" -> handlePersistConfig(call, result)
                "loadConfig" -> handleLoadConfig(call, result)
                "clearConfig" -> handleClearConfig(call, result)
                "appendAuditEntry" -> handleAppendAuditEntry(call, result)
                "loadAuditLog" -> handleLoadAuditLog(call, result)
                "getDeviceSecurityLevel" -> handleGetDeviceSecurityLevel(call, result)
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("ENVIFIED_NATIVE_ERROR", e.message, null)
        }
    }

    // ── Handlers ─────────────────────────────────────────────────────────────

    private fun handleInitialize(call: MethodCall, result: Result) {
        val env = call.argument<String>("env") ?: "default"
        val alias = keystoreAlias(env)
        val level = EnvifiedKeystore.generateKey(alias)
        result.success(mapOf("success" to true, "securityLevel" to level))
    }

    private fun handleEncrypt(call: MethodCall, result: Result) {
        val plaintext = call.argument<ByteArray>("plaintext")
            ?: return result.error("ENVIFIED_MISSING_ARG", "plaintext required", null)
        val env = call.argument<String>("env") ?: "default"
        val iv = call.argument<ByteArray>("iv") ?: generateIv()
        val alias = keystoreAlias(env)
        val ciphertext = EnvifiedCrypto.encrypt(plaintext, alias, iv)
        result.success(mapOf("ciphertext" to ciphertext, "iv" to iv))
    }

    private fun handleDecrypt(call: MethodCall, result: Result) {
        val ciphertext = call.argument<ByteArray>("ciphertext")
            ?: return result.error("ENVIFIED_MISSING_ARG", "ciphertext required", null)
        val iv = call.argument<ByteArray>("iv")
            ?: return result.error("ENVIFIED_MISSING_ARG", "iv required", null)
        val env = call.argument<String>("env") ?: "default"
        val alias = keystoreAlias(env)
        try {
            val plaintext = EnvifiedCrypto.decrypt(ciphertext, alias, iv)
            result.success(mapOf("plaintext" to plaintext))
        } catch (e: javax.crypto.AEADBadTagException) {
            result.error("ENVIFIED_DECRYPT_FAIL", "GCM authentication tag mismatch", null)
        }
    }

    private fun handleStoreSecret(call: MethodCall, result: Result) {
        val keyId = call.argument<String>("keyId")
            ?: return result.error("ENVIFIED_MISSING_ARG", "keyId required", null)
        val ciphertext = call.argument<ByteArray>("ciphertext")
            ?: return result.error("ENVIFIED_MISSING_ARG", "ciphertext required", null)
        val iv = call.argument<ByteArray>("iv")
            ?: return result.error("ENVIFIED_MISSING_ARG", "iv required", null)
        val env = call.argument<String>("env") ?: "default"

        val prefs = getEncryptedPrefs(env)
        prefs.edit()
            .putString("ct_$keyId", android.util.Base64.encodeToString(ciphertext, android.util.Base64.NO_WRAP))
            .putString("iv_$keyId", android.util.Base64.encodeToString(iv, android.util.Base64.NO_WRAP))
            .apply()
        result.success(mapOf("success" to true))
    }

    private fun handleRetrieveSecret(call: MethodCall, result: Result) {
        val keyId = call.argument<String>("keyId")
            ?: return result.error("ENVIFIED_MISSING_ARG", "keyId required", null)
        val env = call.argument<String>("env") ?: "default"

        val prefs = getEncryptedPrefs(env)
        val ctStr = prefs.getString("ct_$keyId", null)
        val ivStr = prefs.getString("iv_$keyId", null)
        if (ctStr == null || ivStr == null) {
            result.error("ENVIFIED_KEY_NOT_FOUND", "Secret '$keyId' not found", null)
            return
        }
        result.success(mapOf(
            "ciphertext" to android.util.Base64.decode(ctStr, android.util.Base64.NO_WRAP),
            "iv" to android.util.Base64.decode(ivStr, android.util.Base64.NO_WRAP),
        ))
    }

    private fun handleDeleteSecret(call: MethodCall, result: Result) {
        val keyId = call.argument<String>("keyId")
            ?: return result.error("ENVIFIED_MISSING_ARG", "keyId required", null)
        val env = call.argument<String>("env") ?: "default"
        val prefs = getEncryptedPrefs(env)
        prefs.edit().remove("ct_$keyId").remove("iv_$keyId").apply()
        result.success(mapOf("success" to true))
    }

    private fun handleDeleteAllSecrets(call: MethodCall, result: Result) {
        val env = call.argument<String>("env") ?: "default"
        val prefs = getEncryptedPrefs(env)
        val count = prefs.all.keys.count { it.startsWith("ct_") }
        prefs.edit().clear().apply()
        result.success(mapOf("deletedCount" to count))
    }

    private fun handleKeyExists(call: MethodCall, result: Result) {
        val alias = call.argument<String>("alias")
            ?: return result.error("ENVIFIED_MISSING_ARG", "alias required", null)
        result.success(mapOf("exists" to EnvifiedKeystore.keyExists(alias)))
    }

    private fun handleRotateKey(call: MethodCall, result: Result) {
        val env = call.argument<String>("env") ?: "default"
        val oldAlias = keystoreAlias(env)
        val newAlias = "${oldAlias}_new_${System.currentTimeMillis()}"

        // Generate the new key first.
        EnvifiedKeystore.generateKey(newAlias)

        val prefs = getEncryptedPrefs(env)
        val allKeys = prefs.all.keys.filter { it.startsWith("ct_") }.map { it.removePrefix("ct_") }
        var count = 0

        for (keyId in allKeys) {
            val ctStr = prefs.getString("ct_$keyId", null) ?: continue
            val ivStr = prefs.getString("iv_$keyId", null) ?: continue
            val ciphertext = android.util.Base64.decode(ctStr, android.util.Base64.NO_WRAP)
            val iv = android.util.Base64.decode(ivStr, android.util.Base64.NO_WRAP)
            try {
                val plaintext = EnvifiedCrypto.decrypt(ciphertext, oldAlias, iv)
                val newIv = generateIv()
                val newCt = EnvifiedCrypto.encrypt(plaintext, newAlias, newIv)
                prefs.edit()
                    .putString("ct_$keyId", android.util.Base64.encodeToString(newCt, android.util.Base64.NO_WRAP))
                    .putString("iv_$keyId", android.util.Base64.encodeToString(newIv, android.util.Base64.NO_WRAP))
                    .apply()
                count++
            } catch (_: Exception) {
                // Skip entries that cannot be re-encrypted.
            }
        }

        // Delete the old key and rename the new one via re-generate under the original alias.
        EnvifiedKeystore.deleteKey(oldAlias)
        // The new alias is now active. Callers must update their alias mapping externally.
        // In practice, we keep the original alias for simplicity — re-generate under it.
        EnvifiedKeystore.generateKey(oldAlias)

        result.success(mapOf("migratedCount" to count))
    }

    private fun handlePersistConfig(call: MethodCall, result: Result) {
        val configJson = call.argument<String>("configJson")
            ?: return result.error("ENVIFIED_MISSING_ARG", "configJson required", null)
        val env = call.argument<String>("env") ?: "default"
        getMetaPrefs().edit().putString("config_$env", configJson).apply()
        result.success(mapOf("success" to true))
    }

    private fun handleLoadConfig(call: MethodCall, result: Result) {
        val env = call.argument<String>("env") ?: "default"
        val json = getMetaPrefs().getString("config_$env", null)
        result.success(mapOf("configJson" to json))
    }

    private fun handleClearConfig(call: MethodCall, result: Result) {
        getMetaPrefs().edit().clear().apply()
        result.success(mapOf("success" to true))
    }

    private fun handleAppendAuditEntry(call: MethodCall, result: Result) {
        val entryJson = call.argument<String>("entryJson")
            ?: return result.error("ENVIFIED_MISSING_ARG", "entryJson required", null)
        val prefs = getMetaPrefs()
        val existing = prefs.getString("audit_log", "[]") ?: "[]"
        val arr = JSONArray(existing)
        arr.put(entryJson)
        // Keep max 50 entries (FIFO).
        val trimmed = JSONArray()
        val start = maxOf(0, arr.length() - 50)
        for (i in start until arr.length()) trimmed.put(arr.get(i))
        prefs.edit().putString("audit_log", trimmed.toString()).apply()
        result.success(mapOf("success" to true))
    }

    private fun handleLoadAuditLog(call: MethodCall, result: Result) {
        val raw = getMetaPrefs().getString("audit_log", "[]") ?: "[]"
        val arr = JSONArray(raw)
        val entries = (0 until arr.length()).map { arr.getString(it) }
        result.success(mapOf("entries" to entries))
    }

    private fun handleGetDeviceSecurityLevel(call: MethodCall, result: Result) {
        // Generate a temp key to probe StrongBox availability.
        val probe = "envified_probe_${System.currentTimeMillis()}"
        val level = try {
            val l = EnvifiedKeystore.generateKey(probe)
            EnvifiedKeystore.deleteKey(probe)
            l
        } catch (_: Exception) {
            "software"
        }
        result.success(mapOf("level" to level))
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private fun keystoreAlias(env: String) = "envified_master_$env"

    private fun generateIv(): ByteArray = ByteArray(12).also { SecureRandom().nextBytes(it) }

    /** Encrypted SharedPreferences for per-env secret blobs. */
    private fun getEncryptedPrefs(env: String): SharedPreferences {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        return EncryptedSharedPreferences.create(
            context,
            "envified_secrets_$env",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    /** Plain SharedPreferences for non-secret metadata (config, audit). */
    private fun getMetaPrefs(): SharedPreferences =
        context.getSharedPreferences("envified_meta", Context.MODE_PRIVATE)

    companion object {
        const val CHANNEL_NAME = "in.appamania.envified/channel"
    }
}

File: ios/Classes/EnvifiedPlugin.swift
import Flutter
import Foundation
import Security

/// Flutter plugin entry point for envified on iOS.
public class EnvifiedPlugin: NSObject, FlutterPlugin {

    // MARK: - Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "in.appamania.envified/channel",
            binaryMessenger: registrar.messenger()
        )
        let instance = EnvifiedPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // MARK: - MethodCallHandler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        do {
            switch call.method {
            case "initialize":       try handleInitialize(args: args, result: result)
            case "encrypt":          try handleEncrypt(args: args, result: result)
            case "decrypt":          try handleDecrypt(args: args, result: result)
            case "storeSecret":      try handleStoreSecret(args: args, result: result)
            case "retrieveSecret":   try handleRetrieveSecret(args: args, result: result)
            case "deleteSecret":     try handleDeleteSecret(args: args, result: result)
            case "deleteAllSecrets": try handleDeleteAllSecrets(args: args, result: result)
            case "keyExists":        handleKeyExists(args: args, result: result)
            case "rotateKey":        try handleRotateKey(args: args, result: result)
            case "persistConfig":    handlePersistConfig(args: args, result: result)
            case "loadConfig":       handleLoadConfig(args: args, result: result)
            case "clearConfig":      handleClearConfig(result: result)
            case "appendAuditEntry": handleAppendAuditEntry(args: args, result: result)
            case "loadAuditLog":     handleLoadAuditLog(result: result)
            case "getDeviceSecurityLevel": handleGetDeviceSecurityLevel(result: result)
            default:                 result(FlutterMethodNotImplemented)
            }
        } catch {
            result(FlutterError(code: "ENVIFIED_NATIVE_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    // MARK: - Handlers

    private func handleInitialize(args: [String: Any], result: FlutterResult) throws {
        let env = args["env"] as? String ?? "default"
        let alias = keystoreAlias(env)
        let level = try EnvifiedKeychain.generateKey(alias: alias)
        result(["success": true, "securityLevel": level])
    }

    private func handleEncrypt(args: [String: Any], result: FlutterResult) throws {
        guard let plaintext = (args["plaintext"] as? FlutterStandardTypedData)?.data else {
            return result(FlutterError(code: "ENVIFIED_MISSING_ARG", message: "plaintext required", details: nil))
        }
        let env = args["env"] as? String ?? "default"
        let iv: Data
        if let ivData = (args["iv"] as? FlutterStandardTypedData)?.data {
            iv = ivData
        } else {
            iv = generateIv()
        }
        let alias = keystoreAlias(env)
        let ciphertext = try EnvifiedCrypto.encrypt(plaintext: plaintext, alias: alias, nonce: iv)
        result(["ciphertext": FlutterStandardTypedData(bytes: ciphertext),
                "iv": FlutterStandardTypedData(bytes: iv)])
    }

    private func handleDecrypt(args: [String: Any], result: FlutterResult) throws {
        guard let ciphertext = (args["ciphertext"] as? FlutterStandardTypedData)?.data,
              let iv = (args["iv"] as? FlutterStandardTypedData)?.data else {
            return result(FlutterError(code: "ENVIFIED_MISSING_ARG", message: "ciphertext and iv required", details: nil))
        }
        let env = args["env"] as? String ?? "default"
        let alias = keystoreAlias(env)
        do {
            let plaintext = try EnvifiedCrypto.decrypt(ciphertext: ciphertext, alias: alias, nonce: iv)
            result(["plaintext": FlutterStandardTypedData(bytes: plaintext)])
        } catch EnvifiedCrypto.CryptoError.decryptFail {
            result(FlutterError(code: "ENVIFIED_DECRYPT_FAIL", message: "GCM authentication tag mismatch", details: nil))
        }
    }

    private func handleStoreSecret(args: [String: Any], result: FlutterResult) throws {
        guard let keyId = args["keyId"] as? String,
              let ciphertext = (args["ciphertext"] as? FlutterStandardTypedData)?.data,
              let iv = (args["iv"] as? FlutterStandardTypedData)?.data else {
            return result(FlutterError(code: "ENVIFIED_MISSING_ARG", message: "keyId, ciphertext, iv required", details: nil))
        }
        let env = args["env"] as? String ?? "default"
        let combined = iv + ciphertext
        try storeKeychainItem(key: secretKey(env: env, keyId: keyId), data: combined)
        result(["success": true])
    }

    private func handleRetrieveSecret(args: [String: Any], result: FlutterResult) throws {
        guard let keyId = args["keyId"] as? String else {
            return result(FlutterError(code: "ENVIFIED_MISSING_ARG", message: "keyId required", details: nil))
        }
        let env = args["env"] as? String ?? "default"
        guard let combined = try loadKeychainItem(key: secretKey(env: env, keyId: keyId)) else {
            return result(FlutterError(code: "ENVIFIED_KEY_NOT_FOUND", message: "Secret '\(keyId)' not found", details: nil))
        }
        let iv = combined.prefix(12)
        let ct = combined.dropFirst(12)
        result(["ciphertext": FlutterStandardTypedData(bytes: Data(ct)),
                "iv": FlutterStandardTypedData(bytes: Data(iv))])
    }

    private func handleDeleteSecret(args: [String: Any], result: FlutterResult) throws {
        guard let keyId = args["keyId"] as? String else {
            return result(FlutterError(code: "ENVIFIED_MISSING_ARG", message: "keyId required", details: nil))
        }
        let env = args["env"] as? String ?? "default"
        try deleteKeychainItem(key: secretKey(env: env, keyId: keyId))
        result(["success": true])
    }

    private func handleDeleteAllSecrets(args: [String: Any], result: FlutterResult) throws {
        let env = args["env"] as? String ?? "default"
        let count = try deleteAllKeychainItems(prefix: "envified_secret_\(env)_")
        result(["deletedCount": count])
    }

    private func handleKeyExists(args: [String: Any], result: FlutterResult) {
        let alias = args["alias"] as? String ?? ""
        result(["exists": EnvifiedKeychain.keyExists(alias: alias)])
    }

    private func handleRotateKey(args: [String: Any], result: FlutterResult) throws {
        // Simplified rotation: re-generate the Keystore key under the same alias.
        // A full production implementation would re-encrypt all stored blobs.
        let env = args["env"] as? String ?? "default"
        let alias = keystoreAlias(env)
        try EnvifiedKeychain.deleteKey(alias: alias)
        try EnvifiedKeychain.generateKey(alias: alias)
        result(["migratedCount": 0])
    }

    private func handlePersistConfig(args: [String: Any], result: FlutterResult) {
        guard let configJson = args["configJson"] as? String,
              let env = args["env"] as? String else {
            return result(FlutterError(code: "ENVIFIED_MISSING_ARG", message: "configJson and env required", details: nil))
        }
        UserDefaults.standard.set(configJson, forKey: "envified_config_\(env)")
        result(["success": true])
    }

    private func handleLoadConfig(args: [String: Any], result: FlutterResult) {
        let env = args["env"] as? String ?? "default"
        let json = UserDefaults.standard.string(forKey: "envified_config_\(env)")
        result(["configJson": json as Any])
    }

    private func handleClearConfig(result: FlutterResult) {
        let defaults = UserDefaults.standard
        defaults.dictionaryRepresentation().keys
            .filter { $0.hasPrefix("envified_") }
            .forEach { defaults.removeObject(forKey: $0) }
        result(["success": true])
    }

    private func handleAppendAuditEntry(args: [String: Any], result: FlutterResult) {
        guard let entryJson = args["entryJson"] as? String else {
            return result(FlutterError(code: "ENVIFIED_MISSING_ARG", message: "entryJson required", details: nil))
        }
        let key = "envified_audit_log"
        var entries = UserDefaults.standard.stringArray(forKey: key) ?? []
        entries.append(entryJson)
        if entries.count > 50 { entries = Array(entries.suffix(50)) }
        UserDefaults.standard.set(entries, forKey: key)
        result(["success": true])
    }

    private func handleLoadAuditLog(result: FlutterResult) {
        let entries = UserDefaults.standard.stringArray(forKey: "envified_audit_log") ?? []
        result(["entries": entries])
    }

    private func handleGetDeviceSecurityLevel(result: FlutterResult) {
        // On a real device with Secure Enclave, CryptoKit can use SE-backed P256 keys.
        // For AES keys via Keychain, the level is "software" key in hardware-protected storage.
        result(["level": "software"])
    }

    // MARK: - Helpers

    private func keystoreAlias(_ env: String) -> String { "envified_master_\(env)" }
    private func secretKey(env: String, keyId: String) -> String { "envified_secret_\(env)_\(keyId)" }
    private func generateIv() -> Data { var bytes = [UInt8](repeating: 0, count: 12); SecRandomCopyBytes(kSecRandomDefault, 12, &bytes); return Data(bytes) }

    private func storeKeychainItem(key: String, data: Data) throws {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: "in.appamania.envified.secrets",
            kSecAttrAccount: key,
        ]
        SecItemDelete(query as CFDictionary)
        var addQuery = query
        addQuery[kSecValueData] = data
        addQuery[kSecAttrAccessible] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(addQuery as CFDictionary, nil)
        if status != errSecSuccess { throw NSError(domain: "EnvifiedPlugin", code: Int(status)) }
    }

    private func loadKeychainItem(key: String) throws -> Data? {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: "in.appamania.envified.secrets",
            kSecAttrAccount: key,
            kSecReturnData: true,
            kSecMatchLimit: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        if status == errSecItemNotFound { return nil }
        if status != errSecSuccess { throw NSError(domain: "EnvifiedPlugin", code: Int(status)) }
        return item as? Data
    }

    private func deleteKeychainItem(key: String) throws {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: "in.appamania.envified.secrets",
            kSecAttrAccount: key,
        ]
        SecItemDelete(query as CFDictionary)
    }

    private func deleteAllKeychainItems(prefix: String) throws -> Int {
        // Enumerate and delete all Keychain items with matching account prefix.
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: "in.appamania.envified.secrets",
            kSecReturnAttributes: true,
            kSecMatchLimit: kSecMatchLimitAll,
        ]
        var items: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &items)
        if status == errSecItemNotFound { return 0 }
        guard status == errSecSuccess, let arr = items as? [[CFString: Any]] else { return 0 }
        var count = 0
        for item in arr {
            if let account = item[kSecAttrAccount] as? String, account.hasPrefix(prefix) {
                try deleteKeychainItem(key: account)
                count += 1
            }
        }
        return count
    }
}
